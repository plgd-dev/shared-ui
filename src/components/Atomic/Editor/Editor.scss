@import '../../../common/styles/colors';

.editor {
  position: relative;

  .jsoneditor {
    border-color: $secondary_20;

    .ace-jsoneditor .ace_marker-layer .ace_active-line {
      background-color: rgba($primary_100, 0.1);
    }

    div.jsoneditor-outer {
      padding: 0 !important;
    }
  }
}
