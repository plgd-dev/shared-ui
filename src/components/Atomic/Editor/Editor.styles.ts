import { css } from '@emotion/react'
import { colors } from '../_utils/colors'

export const editor = css`
    &.disabled:after {
        //display: none;
        background: rgba(255, 255, 255, 0);
    }
    .jsoneditor {
        border: 1px solid #000028;
        border-radius: 8px;
        overflow: hidden;

        .ace-jsoneditor {
            .ace_scroller {
                background: #000028;
            }
            .ace_gutter {
                background: #000028;
                //             //padding: 16px 0;
                //
                .ace_gutter-cell {
                    //font-family: 'Poppins', sans-serif;
                    //font-style: normal;
                    //font-weight: 400;
                    //font-size: 14px;
                    //                 //line-height: 22px;
                    //                 text-align: center;
                    color: ${colors.neutral600};
                    padding: 0 16px;
                    //                 display: flex;
                    //                 align-items: center;
                }
            }

            .ace_gutter-active-line {
                background: #000028;
            }

            .ace_bracket,
            .ace_punctuation,
            .ace_operator,
            .ace_string,
            .ace_constant,
            .ace_paren,
            .ace_variable {
                //font-family: 'Poppins', sans-serif;
                //font-style: normal;
                //font-weight: 400;
                //             font-size: 14px;
                //             //line-height: 22px;
                color: ${colors.neutral600};
            }

            .ace_constant {
                &.ace_boolean {
                    color: ${colors.yellow};
                }

                &.ace_numeric {
                    color: ${colors.red};
                }
            }

            .ace_string {
                color: ${colors.green};
            }
        }
    }
`
