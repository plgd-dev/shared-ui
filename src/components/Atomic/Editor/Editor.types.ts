import { CSSProperties, Ref } from 'react'
import { editorModes } from './constants'

export type EditorModeType = (typeof editorModes)[keyof typeof editorModes]

export type Props = {
    autofocus?: boolean
    className?: boolean
    containerRef?: (ref: Ref<any>) => void
    disabled?: boolean
    editorRef?: (ref: Ref<any> | null) => void
    height?: string
    json: string | [] | object
    mode?: EditorModeType
    onChange?: (json: any) => void
    onError?: (error: any) => void
    onResize?: (width: number, height: number, callback: () => void) => void
    onViewChange?: (fullview: boolean) => void
    schema?: [] | object
    style?: CSSProperties
    width?: string
    i18n?: {
        viewText?: string
    }
}

export const defaultProps = {
    height: '300px',
    mode: editorModes.CODE,
    width: '100%',
}
