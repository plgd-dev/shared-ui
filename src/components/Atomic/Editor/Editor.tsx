// @ts-nocheck
import React, { forwardRef, MutableRefObject, useCallback, useEffect, useImperativeHandle, useRef, useState } from 'react'
import classNames from 'classnames'
import JSONEditor from 'jsoneditor'
import 'jsoneditor/dist/jsoneditor.css'
import isFunction from 'lodash/isFunction'

import { defaultProps, EditorRefType, Props } from './Editor.types'
import * as styles from './Editor.styles'
import IconShowPassword from '../Icon/components/IconShowPassword'

const Editor = forwardRef<EditorRefType, Props>((props, ref) => {
    const {
        autofocus,
        className,
        containerRef,
        dataTestId,
        disabled,
        editorRef,
        height: heightProp,
        i18n,
        json,
        mode,
        onBlur,
        onChange,
        onError,
        onResize,
        onValidation,
        onViewChange,
        schema,
        style,
        width,
        ...rest
    } = { ...defaultProps, ...props }

    const [height, setHeight] = useState(heightProp)

    const container = useRef()
    const jsonEditor = useRef()
    const resizeObserver = useRef()

    const handleContainerRef = (node: any) => {
        if (containerRef) {
            containerRef(node)
        }

        container.current = node
    }

    const onChangeText = (value: any) => {
        const FALSY_VALUES = ['', 'null', 'false', 'undefined']
        const TRUE_VALUES = ['true']

        let returnValue = value

        if (!value || FALSY_VALUES.includes(value)) {
            returnValue = false
        } else if (!value || TRUE_VALUES.includes(value)) {
            returnValue = true
        } else if (!isNaN(Number(value))) {
            returnValue = Number(value)
        } else if (value === 'null') {
            returnValue = null
        }

        if (onChange) {
            onChange(returnValue)
        }
    }

    const onValidationError = (error: any) => {
        if (onError) {
            onError(error)
        }
    }

    const handleEditorRef = (editor: MutableRefObject<undefined> | null) => {
        if (editorRef) {
            editorRef(editor)
        }
    }

    // Triggers the resize method on the editor when resizing its parent container.
    const handleResize = (entries: any) => {
        const { width, height } = entries?.[0]?.contentRect || {}

        if (isFunction(onResize) && onResize) {
            onResize(width, height, () => {
                // @ts-ignore
                jsonEditor?.current?.aceEditor?.resize?.()
            })
        }
    }

    useEffect(() => {
        if (heightProp !== height) {
            setHeight(heightProp)
        }
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [heightProp])

    useEffect(() => {
        const options = {
            mode,
            mainMenuBar: false,
            statusBar: false,
            onChangeText,
            onValidationError: onValidationError,
            onValidate: onValidation,
            schema,
        }

        jsonEditor.current = new JSONEditor(container.current, options)
        if (typeof json === 'object') {
            // @ts-ignore
            jsonEditor?.current?.set(json)
        } else if (typeof json === 'string') {
            // @ts-ignore
            jsonEditor?.current?.setText(json)
        }

        if (autofocus) {
            // @ts-ignore
            jsonEditor?.current?.focus()
        }

        const numberOfLines = jsonEditor.current.aceEditor.env.document.doc.$lines.length || 0
        const lineHeight = jsonEditor.current.aceEditor.renderer.lineHeight || 0

        heightProp === undefined && setHeight(Math.max(numberOfLines * lineHeight + 5, 300))

        // lineHeight
        jsonEditor.current.aceEditor.setOptions({
            minLines: 1000,
            fontSize: 14,
            readOnly: disabled,
            // highlightGutterLine: false,
            // showLineNumbers: false,
        })

        if (typeof ResizeObserver === 'function' && container.current) {
            resizeObserver.current = new ResizeObserver(handleResize)
            resizeObserver.current.observe(container.current)
        }

        handleEditorRef(jsonEditor)

        return () => {
            if (jsonEditor && jsonEditor.current) {
                handleEditorRef(null)
                // @ts-ignore
                jsonEditor?.current?.destroy()
            }

            if (typeof ResizeObserver === 'function' && container && container.current) {
                // @ts-ignore
                resizeObserver?.current?.unobserve(container.current)
            }
        }
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [])

    const handleBlur = useCallback(() => {
        isFunction(onBlur) && onBlur(jsonEditor?.current?.getText())
    }, [onBlur])

    const setValue = useCallback(
        (value) => {
            if (typeof json === 'object') {
                // @ts-ignore
                jsonEditor?.current?.set(value)
            } else if (typeof json === 'string') {
                // @ts-ignore
                jsonEditor?.current?.setText(value)
            }
        },
        [json]
    )

    useImperativeHandle(ref, () => ({
        setValue,
    }))

    return (
        <div
            {...rest}
            className={classNames(className, 'editor', {
                disabled,
                resize: !!ResizeObserver,
            })}
            css={styles.editor}
            data-test-id={dataTestId}
            onBlur={handleBlur}
            ref={(ref) => handleContainerRef(ref)}
            style={{ ...style, width, height }}
        >
            {isFunction(onViewChange) && i18n?.viewText && (
                <a
                    css={styles.fullSizeBtn}
                    data-test-id={dataTestId?.concat('-view-button')}
                    href='#'
                    onClick={(e) => {
                        e.preventDefault()
                        onViewChange()
                    }}
                >
                    <IconShowPassword /> <span css={styles.text}>{i18n?.viewText}</span>
                </a>
            )}

            <input css={styles.testInput} data-test-id={dataTestId?.concat('-input')} onChange={(e) => setValue(JSON.parse(e.target.value))} type='text' />
        </div>
    )
})

Editor.displayName = 'Editor'

export default Editor
