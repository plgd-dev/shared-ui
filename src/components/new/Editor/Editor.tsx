// @ts-nocheck
import React, { FC, MutableRefObject, useEffect, useRef } from 'react'
import { defaultProps, Props } from './Editor.types'
import * as styles from './Editor.styles'
import classNames from 'classnames'
import JSONEditor from 'jsoneditor'
import 'jsoneditor/dist/jsoneditor.css'
import isFunction from 'lodash/isFunction'

const Editor: FC<Props> = (props) => {
    const { autofocus, className, editorRef, containerRef, onChange, onError, json, schema, height, width, style, onResize, disabled, mode, ...rest } = props

    const container = useRef()
    const jsonEditor = useRef()
    const resizeObserver = useRef()

    const handleContainerRef = (node: any) => {
        if (containerRef) {
            containerRef(node)
        }

        container.current = node
    }

    const onChangeText = (json: any) => {
        if (onChange) {
            onChange(json)
        }
    }

    const onValidationError = (error: any) => {
        if (onError) {
            onError(error)
        }
    }

    const handleEditorRef = (editor: MutableRefObject<undefined> | null) => {
        if (editorRef) {
            editorRef(editor)
        }
    }

    // Triggers the resize method on the editor when resizing its parent container.
    const handleResize = (entries: any) => {
        const { width, height } = entries?.[0]?.contentRect || {}

        if (isFunction(onResize) && onResize) {
            onResize(width, height, () => {
                // @ts-ignore
                jsoneditor?.current?.aceEditor?.resize?.()
            })
        }
    }

    useEffect(() => {
        // setTimeout(() => {
        const options = {
            mode,
            mainMenuBar: false,
            statusBar: false,
            onChangeText,
            onValidationError: onValidationError,
            schema,
            disabled: true,
        }

        jsonEditor.current = new JSONEditor(container.current, options)
        if (typeof json === 'object') {
            // @ts-ignore
            jsonEditor?.current?.set(json)
        } else if (typeof json === 'string') {
            // @ts-ignore
            jsonEditor?.current?.setText(json)
        }

        if (autofocus) {
            // @ts-ignore
            jsonEditor?.current?.focus()
        }

        // lineHeight
        jsonEditor.current.aceEditor.setOptions({
            minLines: 1000,
            fontSize: 14,
            // highlightGutterLine: false,
        })

        if (typeof ResizeObserver === 'function' && container.current) {
            resizeObserver.current = new ResizeObserver(handleResize)
            resizeObserver.current.observe(container.current)
        }

        handleEditorRef(jsonEditor)
        // }, 1)

        return () => {
            if (jsonEditor && jsonEditor.current) {
                handleEditorRef(null)
                // @ts-ignore
                jsonEditor?.current?.destroy()
            }

            if (typeof ResizeObserver === 'function' && container && container.current) {
                // @ts-ignore
                resizeObserver?.current?.unobserve(container.current)
            }
        }
    }, [])

    return (
        <div
            {...rest}
            className={classNames(className, 'editor', {
                disabled,
                resize: !!ResizeObserver,
            })}
            css={styles.editor}
            ref={(ref) => handleContainerRef(ref)}
            style={{ ...style, width, height }}
        />
    )
}

Editor.displayName = 'Editor'
Editor.defaultProps = defaultProps

export default Editor
