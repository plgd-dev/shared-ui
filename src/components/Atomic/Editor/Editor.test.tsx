import { render } from '@testing-library/react'
import Editor from './Editor'

describe('<ConditionalWrapper>', () => {
    it('render correctly - snapshot', () => {
        const { asFragment } = render(
            <div>
                <Editor json={{}} />
            </div>
        )

        expect(asFragment()).toMatchSnapshot()
    })
})
