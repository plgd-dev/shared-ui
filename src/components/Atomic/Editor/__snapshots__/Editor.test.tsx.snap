// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ConditionalWrapper> render correctly - snapshot 1`] = `
<DocumentFragment>
  .emotion-0 {
  position: relative;
}

.emotion-0.disabled:after {
  background: rgba(255, 255, 255, 0);
}

.emotion-0 .jsoneditor {
  border: 1px solid;
  border-radius: 8px;
  overflow: hidden;
}

.emotion-0 .jsoneditor .ace-jsoneditor .ace_gutter .ace_gutter-cell {
  padding: 0 16px;
}

.emotion-0 .jsoneditor .ace-jsoneditor .ace_indent-guide {
  background: none;
}

<div>
    <div
      class="editor resize emotion-0"
      style="width: 100%; height: 300px;"
    >
      <div
        class="jsoneditor jsoneditor-mode-code"
      >
        <div
          class="jsoneditor-outer"
          style="margin-bottom: -1px; padding-bottom: 1px;"
        >
          <div
            class=" ace_editor ace_hidpi ace-jsoneditor"
            style="height: 100%; width: 100%; font-size: 14px;"
          >
            <textarea
              aria-autocomplete="both"
              aria-haspopup="false"
              autocapitalize="off"
              autocorrect="off"
              class="ace_text-input"
              role="textbox"
              spellcheck="false"
              style="opacity: 0; font-size: 1px;"
              wrap="off"
            />
            <div
              aria-hidden="true"
              class="ace_gutter"
            >
              <div
                class="ace_layer ace_gutter-layer ace_folding-enabled"
                style="height: 1000000px;"
              />
            </div>
            <div
              class="ace_scroller"
              style="line-height: 0px;"
            >
              <div
                class="ace_content"
              >
                <div
                  class="ace_layer ace_print-margin-layer"
                >
                  <div
                    class="ace_print-margin"
                    style="left: 4px; visibility: hidden;"
                  />
                </div>
                <div
                  class="ace_layer ace_marker-layer"
                />
                <div
                  class="ace_layer ace_text-layer"
                  style="height: 1000000px; margin: 0px 4px;"
                />
                <div
                  class="ace_layer ace_marker-layer"
                />
                <div
                  class="ace_layer ace_cursor-layer ace_hidden-cursors"
                >
                  <div
                    class="ace_cursor"
                  />
                </div>
              </div>
            </div>
            <div
              class="ace_scrollbar ace_scrollbar-v"
              style="display: none; width: 20px;"
            >
              <div
                class="ace_scrollbar-inner"
                style="width: 20px;"
              >
                 
              </div>
            </div>
            <div
              class="ace_scrollbar ace_scrollbar-h"
              style="display: none; height: 20px;"
            >
              <div
                class="ace_scrollbar-inner"
                style="height: 20px;"
              >
                 
              </div>
            </div>
            <div
              style="height: auto; width: auto; top: 0px; left: 0px; visibility: hidden; position: absolute; white-space: pre; overflow: hidden;"
            >
              <div
                style="height: auto; width: auto; top: 0px; left: 0px; visibility: hidden; position: absolute; white-space: pre; overflow: visible;"
              />
              <div
                style="height: auto; width: auto; top: 0px; left: 0px; visibility: hidden; position: absolute; white-space: pre; overflow: visible;"
              >
                XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
              </div>
            </div>
          </div>
        </div>
        <div
          class="jsoneditor-validation-errors-container"
        >
          <div
            class="jsoneditor-additional-errors fadein"
            style="display: none;"
          >
            Scroll for more ▿
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
